chain/beacon/sync_manager.go

```diff
@@ -476,65 +550,48 @@ func SyncChain(l log.Logger, store CallbackStore, req SyncRequest, stream SyncSt
 	}
 
 	send := func(b *chain.Beacon) error {
-		packet := beaconToProto(b)
-		packet.Metadata = &common.Metadata{BeaconID: beaconID}
-		err := stream.Send(packet)
+		err := stream.Send(beaconToProto(b, beaconID))
 		if err != nil {
 			logger.Debugw("", "syncer", "streaming_send", "err", err)
 		}
 		return err
 	}
 
-	// we know that last.Round >= fromRound from the above if
 	if fromRound != 0 {
-		// TODO (dlsniper): During the loop below, we can receive new data
-		//  which may not be observed as the callback is added after the loop ends.
-		//  Investigate if how the storage view updates while the cursor runs.
-
-		// first sync up from the store itself
-		err = store.Cursor(ctx, func(ctx context.Context, c chain.Cursor) error {
-			bb, err := c.Seek(ctx, fromRound)
-			for ; bb != nil; bb, err = c.Next(ctx) {
-				// This is needed since send will use a pointer and could result in pointer reassignment
-				bb := bb
-				if err != nil {
-					return err
-				}
-				// Force send the correct
-				if err := send(bb); err != nil {
-					logger.Debugw("Error while sending beacon", "syncer", "cursor_seek")
-					return err
-				}
-			}
-			return err
-		})
+		// first we sync up
+		err = SyncChain(l, store, req, stream)
 		if err != nil {
-			// We always have ErrNoBeaconStored returned as last value
-			// so let's ignore it and not send it back to the client
-			if !errors.Is(err, chainerrors.ErrNoBeaconStored) {
-				return err
-			}
+			logger.Errorw("Error while processing SyncChain request", "err", err)
+			return err
 		}
 	}
 
+	select {
+	case <-ctx.Done():
+		logger.Debugw("Context done.")
+		return ctx.Err()
+	default:
+		logger.Debugw("Adding callback to sync new beacons", "id", id)
+	}
 	// Register a callback to process all new incoming beacons until an error happens.
 	// The callback happens in a separate goroutine.
 	errChan := make(chan error)
 	store.AddCallback(id, func(b *chain.Beacon) {
 		if err := send(b); err != nil {
-			logger.Debugw("Error while sending beacon", "syncer", "callback")
+			logger.Debugw("Error while sending beacon", "syncer", "callback", "err", err)
 			store.RemoveCallback(id)
 			errChan <- err
 		}
 	})
 
-	defer store.RemoveCallback(id)
-
 	// Wait until the request cancels or until an error happens in the callback.
 	select {
 	case <-ctx.Done():
```

Comment:
won't this still hang if `SyncChain` completes with `nil` error but nobody ever cancels the context?

Reply:
Nope, since it's not added before the SyncChain call